src/backend/distributed/utils/resource_lock.c

```diff
@@ -122,6 +124,90 @@ lock_shard_resources(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * LockShardListResourcesOnFirstWorker acquires the resource locks for the specified
+ * shards on the first worker. Note that the function does not sort the shard list,
+ * therefore the caller should sort the shard list in order to avoid deadlocks.
+ */
+void
+LockShardListResourcesOnFirstWorker(LOCKMODE lockmode, List *shardIntervalList)
+{
+	List *workerNodeList = ActivePrimaryNodeList();
+	WorkerNode *firstWorkerNode = NULL;
+	StringInfo lockCommand = makeStringInfo();
+	ListCell *shardIntervalCell = NULL;
+	int processedShardPlacementCount = 0;
+	int totalShardPlacementCount = list_length(shardIntervalList);
+
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodesByGroupId);
```

Comment:
Can we move this logic into a separate function like `IsFirstWorker()` and do

```C
if (IsFirstWorker())
   return;
```

Actually, I think we should probably do

```C
if firstWorker
   LockShardResourceList(..)
  return SHARD_RESOURCES_LOCKED_LOCALLY

...
return SHARD_RESOURCES_LOCKED_REMOTELY
```

Because otherwise the name of the function won't match what its doing. We can later use the return value to decide on calling `LockShardResource` such as : 

```C


	if (list_length(referenceShardIntervalList) != 0)
	{
		if (LockShardListResourcesOnFirstWorker(lockMode, referenceShardIntervalList) != SHARD_RESOURCES_LOCKED_LOCALLY)
		{
			LockShardResourceList();
		}
	}
```

We could even consider not retuning any value and call `LockShardResourceList();` with the cost re-locking the same shard intervals. However, I think I'd prefer returning a value to signal the caller whether we've locked locally or remotely.